fe/estimator_abfe.py

```diff
@@ -300,53 +241,69 @@ def _deltaG(model, sys_params) -> Tuple[Tuple[float, List], np.array]:
-        # (ytz): figure out what to do with stddev(du_dl) later
-        # print(f"{model.prefix} index {lambda_idx} lambda {lambda_window:.5f} <du/dl> {np.mean(result.du_dls):.5f} med(du/dl) {np.median(result.du_dls):.5f}  o(du/dl) {np.std(result.du_dls):.5f}")
-        mean_du_dls.append(np.mean(result.du_dls))
+    U_knk = []
+    N_k = []
+    for lambda_idx, (lambda_window, result) in enumerate(zip(model.lambda_schedule, sim_results)):
+        U_knk.append(result.lambda_us)
+        N_k.append(len(result.lambda_us)) # number of frames
         all_grads.append(result.du_dps)
 
-    tibar_dG = 0
+    U_knk = np.array(U_knk)
+
     bar_dG = 0
     bar_dG_err = 0
 
+    delta_Us = extract_delta_Us_from_U_knk(U_knk)
+
     for lambda_idx in range(len(model.lambda_schedule) - 1):
-        # tibar
-        lamb_start = model.lambda_schedule[lambda_idx]
-        lamb_end = model.lambda_schedule[lambda_idx+1]
-        delta_lamb = lamb_end - lamb_start
-        fwd_work_approx = ti_results[lambda_idx].du_dls*delta_lamb
-        rev_work_approx = -ti_results[lambda_idx+1].du_dls*delta_lamb
-        tibar, ti_bar_err = pymbar.BAR(model.beta*fwd_work_approx, model.beta*rev_work_approx)
-        tibar_overlap = endpoint_correction.overlap_from_cdf(fwd_work_approx, rev_work_approx)
-        tibar_dG += tibar/model.beta
-
-        # exact_bar
-        fwd_work_exact = model.beta*ti_results[lambda_idx].right_dus
-        rev_work_exact = model.beta*ti_results[lambda_idx+1].left_dus
-
-        dG_exact, exact_bar_err = pymbar.BAR(fwd_work_exact, rev_work_exact)
+
+        fwd_delta_u = model.beta*delta_Us[lambda_idx][0]
+        rev_delta_u = model.beta*delta_Us[lambda_idx][1]
+
+        dG_exact, exact_bar_err = pymbar.BAR(fwd_delta_u, rev_delta_u)
         bar_dG += dG_exact/model.beta
-        exact_bar_overlap = endpoint_correction.overlap_from_cdf(fwd_work_exact, rev_work_exact)
+        exact_bar_overlap = endpoint_correction.overlap_from_cdf(fwd_delta_u, rev_delta_u)
 
-        # probably off by a factor of two
+        # probably off by a factor of two since we re-use samples.
         bar_dG_err += (exact_bar_err/model.beta)**2
 
+        lamb_start = model.lambda_schedule[lambda_idx]
+        lamb_end = model.lambda_schedule[lambda_idx+1]
+
         print(f"{model.prefix}_BAR: lambda {lamb_start:.3f} -> {lamb_end:.3f} dG: {dG_exact/model.beta:.3f} dG_err: {exact_bar_err/model.beta:.3f} overlap: {exact_bar_overlap:.3f}")
 
+    # for MBAR we need to sanitize the energies
+    cleans_U_knks = [] # [K, F, K]
+    for lambda_idx, full_us in enumerate(U_knk):
+        cleans_U_knks.append(sanitize_energies(full_us, lambda_idx))
+
+    K = len(model.lambda_schedule)
+    cleans_U_knks = np.array(cleans_U_knks) # [K, F, K]
+    U_kn = np.reshape(cleans_U_knks, (-1, K)).transpose() # [K, F*K]
+    U_kn = U_kn*model.beta
+
+    np.save(model.prefix+"_U_kn.npy", U_kn)
+
+    mbar = pymbar.MBAR(U_kn, N_k)
```

Comment:
Preferred convention: use lower-case `u` for reduced potential (unitless), upper-case `U` for potential energy (in energy units). If adopted, could reduce ambiguity about whether `U_kn.npy` was saved before or after multiplication by `beta`.

```suggestion
    u_kn = U_kn*model.beta

    np.save(model.prefix+"_U_kn.npy", U_kn)

    mbar = pymbar.MBAR(u_kn, N_k)
```